data/transactions/verify/txn.go

```diff
@@ -502,3 +530,382 @@ func (w *worksetBuilder) next() (txnGroups [][]transactions.SignedTxn) {
+	ctx              context.Context
+	cache            VerifiedTransactionCache
+	activeLoopWg     sync.WaitGroup
+	nbw              *NewBlockWatcher
+	ledger           logic.LedgerForSignature
+	droppedFromPool  *metrics.Counter
+}
+
+// NewBlockWatcher is a struct used to provide a new block header to the
+// stream verifier
+type NewBlockWatcher struct {
+	blkHeader bookkeeping.BlockHeader
+	mu        deadlock.RWMutex
+}
+
+// MakeNewBlockWatcher construct a new block watcher with the initial blkHdr
+func MakeNewBlockWatcher(blkHdr bookkeeping.BlockHeader) (nbw *NewBlockWatcher) {
+	nbw = &NewBlockWatcher{
+		blkHeader: blkHdr,
+	}
+	return nbw
+}
+
+// OnNewBlock implements the interface to subscribe to new block notifications from the ledger
+func (nbw *NewBlockWatcher) OnNewBlock(block bookkeeping.Block, delta ledgercore.StateDelta) {
+	if nbw.blkHeader.Round >= block.BlockHeader.Round {
+		return
+	}
+	nbw.mu.Lock()
+	defer nbw.mu.Unlock()
+	nbw.blkHeader = block.BlockHeader
+}
+
+func (nbw *NewBlockWatcher) getBlockHeader() (bh *bookkeeping.BlockHeader) {
+	nbw.mu.RLock()
+	defer nbw.mu.RUnlock()
+	return &nbw.blkHeader
+}
+
+type batchLoad struct {
+	txnGroups             [][]transactions.SignedTxn
+	groupCtxs             []*GroupContext
+	elementBacklogMessage []interface{}
+	messagesForTxn        []int
+}
+
+func makeBatchLoad() (bl batchLoad) {
+	bl.txnGroups = make([][]transactions.SignedTxn, 0)
+	bl.groupCtxs = make([]*GroupContext, 0)
+	bl.elementBacklogMessage = make([]interface{}, 0)
+	bl.messagesForTxn = make([]int, 0)
+	return bl
+}
+
+func (bl *batchLoad) addLoad(txngrp []transactions.SignedTxn, gctx *GroupContext, backlogMsg interface{}, numBatchableSigs int) {
+	bl.txnGroups = append(bl.txnGroups, txngrp)
+	bl.groupCtxs = append(bl.groupCtxs, gctx)
+	bl.elementBacklogMessage = append(bl.elementBacklogMessage, backlogMsg)
+	bl.messagesForTxn = append(bl.messagesForTxn, numBatchableSigs)
+
+}
+
+// MakeStreamVerifier creates a new stream verifier and returns the chans used to send txn groups
+// to it and obtain the txn signature verification result from
+func MakeStreamVerifier(stxnChan <-chan UnverifiedElement, resultChan chan<- VerificationResult,
+	ledger logic.LedgerForSignature, nbw *NewBlockWatcher, verificationPool execpool.BacklogPool,
+	cache VerifiedTransactionCache, droppedFromPool *metrics.Counter) (sv *StreamVerifier) {
+
+	sv = &StreamVerifier{
+		resultChan:       resultChan,
+		stxnChan:         stxnChan,
+		verificationPool: verificationPool,
+		cache:            cache,
+		nbw:              nbw,
+		ledger:           ledger,
+		droppedFromPool:  droppedFromPool,
+	}
+	return sv
+}
+
+// Start is called when the verifier is created and whenever it needs to restart after
+// the ctx is canceled
+func (sv *StreamVerifier) Start(ctx context.Context) {
+	sv.ctx = ctx
+	sv.activeLoopWg.Add(1)
+	go sv.batchingLoop()
+}
+
+func (sv *StreamVerifier) cleanup(pending *[]UnverifiedElement) {
```

Comment:
since it is a slice, you don't need to pass it by pointer.

Reply:
Yes, but this is called from `defer sv.cleanup(&uelts)` and I need the cleanup to be called on the latest value of `uelts` not the one when the `defer` statement is executed.

Reply:
yes. I prefer the second option!